test/integration/integration_test.go

```diff
@@ -0,0 +1,208 @@
+	Value string `json:"value"`
+}
+
+func newLabClient() (*labClient, error) {
+	secretS := os.Getenv("clientSecret")
+	secret, err := msal.CreateClientCredentialFromSecret(secretS)
+	if err != nil {
+		return nil, err
+	}
+	options := msal.DefaultConfidentialClientApplicationOptions()
+	clientID := os.Getenv("clientId")
+	options.Authority = microsoftAuthority
+	app, err := msal.NewConfidentialClientApplication(clientID, secret, &options)
+	if err != nil {
+		return nil, err
+	}
+	return &labClient{app: app}, nil
+}
+func (l *labClient) getLabAccessToken() (string, error) {
+	scopes := []string{msIDlabDefaultScope}
+	result, err := l.app.AcquireTokenSilent(context.Background(), scopes, nil)
+	if err != nil {
+		result, err = l.app.AcquireTokenByClientCredential(context.Background(), scopes)
+		if err != nil {
+			return "", err
+		}
+	}
+	return result.GetAccessToken(), nil
+}
+
+func (l *labClient) getUser(query map[string]string) (user, error) {
+	accessToken, err := l.getLabAccessToken()
+	if err != nil {
+		return user{}, err
+	}
+	responseBody, err := httpRequest("https://msidlab.com/api/user", query, accessToken)
+	if err != nil {
+		return user{}, err
+	}
+	var users []user
+	err = json.Unmarshal(responseBody, &users)
+	if err != nil {
+		return user{}, err
+	}
+	if len(users) == 0 {
+		return user{}, errors.New("No user found")
+	}
+	user := users[0]
+	user.Password, err = l.getSecret(map[string]string{"Secret": user.LabName})
+	if err != nil {
+		return user, err
+	}
+	return user, nil
+}
+
+func (l *labClient) getSecret(query map[string]string) (string, error) {
+	accessToken, err := l.getLabAccessToken()
+	if err != nil {
+		return "", err
+	}
+	responseBody, err := httpRequest("https://msidlab.com/api/LabUserSecret", query, accessToken)
+	if err != nil {
+		return "", err
+	}
+	var secret secret
+	err = json.Unmarshal(responseBody, &secret)
+	if err != nil {
+		return "", err
+	}
+	return secret.Value, nil
+}
+
+// TODO: Add getApp() when needed
+
+// TestConstants
+
+// Scopes.
+const (
+	msIDlabDefaultScope = "https://msidlab.com/.default"
+	graphDefaultScope   = "https://graph.windows.net/.default"
+)
+
+// MicrosoftAuthorityHost is the host authority for Microsoft.
+const microsoftAuthorityHost = "https://login.microsoftonline.com/"
+
+// Authority values.
+const (
+	organizationsAuthority = microsoftAuthorityHost + "organizations/"
+	microsoftAuthority     = microsoftAuthorityHost + "microsoft.onmicrosoft.com"
+	//msIDlabTenantAuthority = microsoftAuthorityHost + "msidlab4.onmicrosoft.com" - Will be needed in the furture
+)
+
+// Tests
+
+func getTestUser(lc *labClient, query map[string]string) user {
+	testUser, err := lc.getUser(query)
+	if err != nil {
+		panic("Failed to get input user. " + err.Error())
+	}
+	return testUser
+}
+
+func TestUsernamePassword(t *testing.T) {
+	labClientInstance, err := newLabClient()
+	if err != nil {
+		panic("Failed to get a lab client. " + err.Error())
+	}
+	tests := map[string]struct {
```

Comment:
This is must improved. But I'm not a fan of the map[string]struct{}, mainly because I like things to execute in a consistent order over time.  This will execute in random order.  Using []struct is more standard.

```go
tests := []struct{
  desc string
  user user
}{
  {"Managed", getTestUser(labClientInstance, map[string]string{"usertype": "cloud"})},
  ...
}
```